### ISSUE TITLE
[Title][Accessibility] Possibility of adding description related to link in Title component

### ISSUE THREAD

**Issue by patrycjakurkowiak** [adobe/aem-core-wcm-components#668]
## Feature Request

**Current behavior**
As an author I would like to add description (like Alt text) to link in Title component. When I try to configure Title component there is only field to fill link (without description for this link).

**Expected behaviour**
Author can add purpose of the link (for instance by some Alt text field).

**Are there alternatives?**
Raised issue is related to 2.4.4 point from WCAG 2.0 accessibility standard. Is there any buisness goal that this component doesn't require 'Alt text' field?

**Environment**
AEM version 6.5.0
Core Components 2.5.0

**Comment by richardhand** [comment]
Hi @patrycjakurkowiak, thanks for the report. alternative text, as in the `alt` attribute isn't valid for anchors. Rather the text content of the anchor is taken as descriptive. It is already possible to edit the title of this anchor in the title component. 

How do you expect this additional descriptive text be presented in the title markup? Do you have an example?

I read through the accompanying understanding document for `2.4.4` of WCAG 2.0 [0] and it seems like the onus is on the component author to ensure that both the context and the provided label for the title meets this criteria.

[0] - https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-refs.html

**Comment by patrycjakurkowiak** [comment]
Hi
Thanks for your answer. I understand that.
I have some concerns related to external links (I am wondering whether there is some business goal to add Title with external link).
Actually, could you introduce some general way of giving information about external links in different components (Title or Teaser components, generally)? I expect that screen reader should read this information and I would like introduce some general solution.
For instance I can add this information in Text component (by Alt text). Probably author should provide such information in some general way (to keep consistency).

**Comment by richardhand** [comment]
> I have some concerns related to external links (I am wondering whether there is some business goal to add Title with external link).

It is already possible to add external links in the Title (and Teaser, and other components) in the `link` edit dialog field.

> Actually, could you introduce some general way of giving information about external links in different components (Title or Teaser components, generally)? I expect that screen reader should read this information and I would like introduce some general solution.
For instance I can add this information in Text component (by Alt text). Probably author should provide such information in some general way (to keep consistency).

I'm afraid I still don't understand the ask here, could you provide: 
• An example of how you are using the text component to provide this description. "Alt text" is confusing to me as it is specific to images or input HTML elements. 
• An example of how you would expect the markup of the Title and Teaser components to look to support this extra information for screen readers.

**Comment by patrycjakurkowiak** [comment]
Hi,
Thanks for your answer.
> It is already possible to add external links in the Title (and Teaser, and other components) in the link edit dialog field.

Yes I know.

For instance, let's assume that I have a link with some title (but it is a link to external service). I expect that screen reader can read information about that user is redirected to another service (if there is some business goal to do it in that way). The same situation we have got in Teaser component. There are no any labels related to links defined as a "call to action".

**Comment by richardhand** [comment]
> For instance, let's assume that I have a link with some title (but it is a link to external service). I expect that screen reader can read information about that user is redirected to another service (if there is some business goal to do it in that way).

As I understand most screen readers are able to read the link href to a user, so that alone would already provide some additional context on external/internal links, if required.

> The same situation we have got in Teaser component. There are no any labels related to links defined as a "call to action".

This isn't really true, you can provide a link text for the actions, which according to 2.4.4 of WCAG 2.0 [0] is all that is required.

> I expect that screen reader can read information about that user is redirected to another service

I think we're still missing an example here of you expect this additional information to be provided in the link markup. Could you please provide a markup example, as requested? And an example of how you are satisfying your requirement in the text component? Thanks!

[0] ```2.4.4 Link Purpose (In Context): The purpose of each link can be determined from the link text alone or from the link text together with its programmatically determined link context, except where the purpose of the link would be ambiguous to users in general. (Level A)```

**Comment by patrycjakurkowiak** [comment]
Hi,
Let's assume that there is some businnes goal to do it and information that link provides to external service should not be visible on page in content.
I expect that example markup should look like below (for Title component):
`<a class="example-class" href="link_to_external_service" title="Link provides to external service">Title</a>`
or
`<a class="example-class" href="link_to_external_service" aria-label="Link provides to external service">Title</a>`

I would like to achieve the same for links in Teaser component defined as a "Call to Action" links.
Could you give me example how it should be set in Teaser component?
I cannot see option to provide this information in markup using title or aria-label.

**Comment by richardhand** [comment]
Hi, thanks for the additional information.

#### `title` Attribute
I think the `title` attribute is more appropriate to your example, as it provides supplementary descriptive information, typically provided in a tooltip. As described in the specification [0]:

> The title attribute represents advisory information for the element, such as would be appropriate for a tooltip.

Unfortunately it seems that the `title` attribute is handled differently for different screen readers and isn't always announced in addition to the anchor `NAME` (derived from text content or aria-label). See the note in the same specification:

> Relying on the title attribute is currently discouraged as many user agents do not expose the attribute in an accessible manner as required by this specification (e.g., requiring a pointing device such as a mouse to cause a tooltip to appear, which excludes keyboard-only users and touch-only users, such as anyone with a modern phone or tablet).

#### `aria-label` Attribute
This is useful in the situation where the text content of the link is ambiguous (I.e. short or not very descriptive) and a more descriptive label would be useful to impaired uses, given that they don't have the full context that a non-impaired user does. An example from the aria guidelines [1]:

```
<a href="taxhike.html" aria-label="Read more about Seminole tax hike">[Read more...]</a>
```

The `aria-label` typically replaces the text content when read out by screen readers, so in your example, only "Link provides to external service" would be announced, not "title". This isn't really useful to an impaired user as you don't have information about the link name. 

#### Adding support in Core Components
I think a case could be made for supporting both the `title` and `aria-label` attributes in the Core Components where relevant. We already have one issue open for `aria-label` support in the Tabs, Navigation and Carousel components, but off the top of my head I think they would also be useful for Button, Teaser Actions, Linked Title, Linked Image.

I agree that a general approach cross component is best.

[0] https://html.spec.whatwg.org/multipage/dom.html#attr-title
[1] https://www.w3.org/TR/WCAG20-TECHS/aria#ARIA8
[2] https://github.com/adobe/aem-core-wcm-components/issues/586

**Comment by nathanmckean** [comment]
aria-label is already available on button. As I think you've hidden, the core issue is inconsistency in accessibility/functionality across components. Where you have a link or linked object, aria-label is useful to provide context when the text is abridged for visual users.

The other issue is around image alt text. Depending on the context of how an image is used, that role of the image might be descriptive (alt text should describe the image), decorative (alt text should be emplty alt="") or functional/linked (alt text should describe the destination). This is not consistency applied across components (e.g. teaser). There is also an a separate issue regarding duplication of links which is raised elsewhere but all of these things are connected.
